Generalize the viewer ID machinery for nonstandard names

bmorris3 opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

The goal of this PR is to allow users to specify custom `jdaviz.Application` configurations, with custom viewer reference names that are _not_ already known to jdaviz (like the familiar ones `"spectrum-viewer"`, `"table-viewer"`, etc.). Most of the effort in this PR went into adapting the existing viewer ID machinery to gracefully handle cases when the viewer's name is not one of the common ones. I've added one new test which checks that a custom configuration with nonstandard names can be initialized.

**Update 10/7/2022**: This PR has grown and evolved a lot since the first commits. I wrote up some [summary and discussion points in this comment below](https://github.com/spacetelescope/jdaviz/pull/1681#issuecomment-1271627712).

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->
### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

bmorris3 commented on jdaviz/tests/test_app.py:
```diff
@@ -1,19 +1,47 @@
+    assert isinstance(plg, GaussianSmooth)
+
+    with pytest.raises(KeyError, match='not found in app'):
+        app.get_tray_item_from_name('imviz-compass')
```
These lines appear to be re-written in the `git diff`, but the code hasn't changed. What _has_ changed is that somehow the newline characters seemed to be `^M`s in this file, and I changed them to standard newlines. 🤷🏻‍♂️

duytnguyendtn replied:
🤷 so long as the tests pass!

pllim commented on jdaviz/app.py:
```diff
@@ -1300,67 +1300,68 @@ def _update_selected_data_items(self, viewer_id, selected_items):
-            selected_items = {si: 'visible' for si in selected_items}
-        # Update the stored selected data items
-        viewer_item['selected_data_items'] = selected_items
+        if viewer is not None:
```
Isn't it less diff if you do this instead? Then you don't have to indent so much.

```python
if viewer is None:
    return
```

pllim commented on jdaviz/app.py:
```diff
@@ -575,7 +575,7 @@ def get_viewer_by_id(self, vid):
         This is useful when reference name is `None`.
         """
-        return self._viewer_store[vid]
+        return self._viewer_store.get(vid)
```
This behavior is now inconsistent with that of `def _viewer_by_reference(self, reference)`. Won't it be too confusing if accessing it one way silently gives `None` but another will still throw exception?

duytnguyendtn replied:
@pllim is this comment still of concern?

pllim replied:
I guess not.

pllim reviewed (commented):
Might need a change log too. I cannot decide if it is bugfix or new feature.

bmorris3 commented:
> Might need a change log too. I cannot decide if it is bugfix or new feature.

I don't know either, it's kind of both?

kecnry commented:
What is the intended scope or use-case of this PR?  I can successfully launch a bare application with a custom config, but then that doesn't have any helper associated with it.  If I create a custom helper by subclassing Specviz, for example, then there are still places in the specviz helper that raise errors because of the lacking `specviz-viewer`.  For example, modifying the test-case here:

```
config = {'settings': {'configuration': 'nonstandard',
                       'data': {'parser': 'specviz-spectrum1d-parser'},
                       'visible': {'menu_bar': False,
                                   'toolbar': True,
                                   'tray': True,
                                   'tab_headers': False},
                       'context': {'notebook': {'max_height': '750px'}}},
          'toolbar': ['g-data-tools', 'g-subset-tools'],
          'tray': ['g-metadata-viewer',
                   'g-plot-options',
                   'g-subset-plugin',
                   'g-gaussian-smooth',
                   'g-model-fitting',
                   'g-unit-conversion',
                   'g-export-plot'],
          'viewer_area': [{'container': 'col',
                           'children': [{'container': 'row',
                                         'viewers': [{'name': 'H',
                                                      'plot': 'specviz-profile-viewer',
                                                      'reference': 'h'},
                                                     {'name': 'K',
                                                      'plot': 'specviz-profile-viewer',
                                                      'reference': 'k'}]}]}]}

from jdaviz import Specviz

class Customviz(Specviz):
    _default_configuration = config
    
viz = Customviz()

from astropy.utils.data import download_file
fn = download_file('https://data.sdss.org/sas/dr14/sdss/spectro/redux/26/spectra/0751/spec-0751-52251-0160.fits', cache=True)

viz.load_spectrum(fn, "myfile", format="SDSS-III/IV spec")
viz.show()
```

fails with `ValueError: Could not identify viewer with reference spectrum-viewer`.  This kind of makes sense, I inherited from specviz, but then excluded the viewer that specviz assumes exists.  I get the same thing if inheriting from `jdaviz.core.helpers.ConfigHelper` instead since `viz.load_data` is still ultimately defaulting to the spectrum parser, which still assumes there is a viewer called "spectrum-viewer".

bmorris3 commented:
Good point @kecnry. To address https://github.com/spacetelescope/jdaviz/pull/1681#issuecomment-1261289614, would it make sense to add a class attribute onto `Specviz` similar to `_default_configuration`, which stores the default name for the viewer? For example, we could do something like: 
```python
class Specviz(ConfigHelper, LineListMixin):
    """Specviz Helper class."""

    _default_configuration = "specviz"
    _default_viewer_reference = "spectrum-viewer"
```
and then rely on `_default_viewer_reference` within `jdaviz/configs/specviz/helper.py`. Thoughts?

kecnry commented:
That could be an option, or maybe an app-level method for `app.get_first_viewer_of_class('SpecvizProfileView')` which can always be used in place of calls to `app.get_viewer('spectrum-viewer')`?  But I also don't want to blow up the scope of this PR if there is a more limited case that this addresses for now, and we can generalize more later down the road.

bmorris3 commented:
@kecnry pointed out above that an error is raised if:
1. a nonstandard configuration is initialized and 
2. there is no available viewer with the viewer reference name `"spectrum-viewer"`

In the latest commits, I've improved the handling in this case. Now if you can load a spectrum with a new optional keyword `viz.load_spectrum(data, viewer_reference_name='nonstandard-name')` to specify which possibly nonstandard viewer into which to load the data. An example looks like this:
```python
config = {'settings': {'configuration': 'nonstandard',
                       'data': {'parser': 'specviz-spectrum1d-parser'},
                       'visible': {'menu_bar': False,
                                   'toolbar': True,
                                   'tray': True,
                                   'tab_headers': False},
                       'context': {'notebook': {'max_height': '750px'}}},
          'toolbar': ['g-data-tools', 'g-subset-tools'],
          'tray': ['g-metadata-viewer',
                   'g-plot-options',
                   'g-subset-plugin',
                   'g-gaussian-smooth',
                   'g-model-fitting',
                   'g-unit-conversion',
                   'g-export-plot'],
          'viewer_area': [{'container': 'col',
                           'children': [{'container': 'row',
                                         'viewers': [{'name': 'H',
                                                      'plot': 'specviz-profile-viewer',
                                                      'reference': 'h'},
                                                     {'name': 'K',
                                                      'plot': 'specviz-profile-viewer',
                                                      'reference': 'k'}]}]}]}

from jdaviz import Specviz

class Customviz(Specviz):
    _default_configuration = config

viz = Customviz()

from astropy.utils.data import download_file
fn = download_file('https://data.sdss.org/sas/dr14/sdss/spectro/redux/26/spectra/0751/spec-0751-52251-0160.fits', cache=True)

# use the new kwarg to specify which viewer to load each spectrum into:
viz.load_spectrum(fn, "h", format="SDSS-III/IV spec", viewer_reference_name='h')
viz.load_spectrum(fn, "k", format="SDSS-III/IV spec", viewer_reference_name='k')
viz.show()
```

I've also added another (more presumptuous?) feature: if there are multiple viewers available in the `Specviz` helper and the user doesn't specify a `viewer_reference_name` in the call to `viz.load_spectrum`, the `Specviz` helper will load the data into the first viewer which does not already have data loaded into it, like this:
```python
# ...using the same config dict as above...
from jdaviz import Specviz

class Customviz(Specviz):
    _default_configuration = config

viz = Customviz()

from astropy.utils.data import download_file
fn = download_file('https://data.sdss.org/sas/dr14/sdss/spectro/redux/26/spectra/0751/spec-0751-52251-0160.fits', cache=True)

viz.load_spectrum(fn, "h", format="SDSS-III/IV spec")  # loads into viewer 'h'
viz.load_spectrum(fn, "k", format="SDSS-III/IV spec")  # loads into viewer 'k'
viz.show()
```

duytnguyendtn commented:
Hmm.... I have some philosophical questions before I give a full review. These are about scope, similar to @kecnry's earlier question:

From my perspective, the helpers are inherently tied to the configuration. Maybe I'm thinking of them differently, but loading a helper (or even subclassing it) with a different config file with it, seems slightly against my understanding of the ethos of the helper. By nature of the helper's goal of providing convenience methods for the user, the helper has to have an understanding of what the configuration is/looks like. For instance, should we allow a user to load a config with an imviz viewer into a specviz-like helper?

Now, that being said, this is ENTIRELY separate from (I think) the good goal to make the specviz helper more robust to tampering from the user. I think the changes Brett introduced after https://github.com/spacetelescope/jdaviz/pull/1681#issuecomment-1261371029 are great, just for a *different* reason. I think this work puts us in a good spot in our direction to make the configs more flexible (e.g. a user deletes the original spectrum viewer, then adds a new one. This PR might prevent everything from breaking)

I really see this PR as a big step towards allowing users to create new configs from scratch, extending `jdaviz` from "that package that has 5 viz tools" to a fully-fledged data analysis platform. But this brings in a scope question from me: what about plugins? At this point, you can create a new config, which may be a worthwhile increment on its own. But doesn't the config file also specify plugins? I think we're still making similar assumptions across the plugins. If this is the case, wouldn't this PR expose those issues if the user tried to include, say, the model fitting plugin in this sample config, then tried to use it?

bmorris3 commented:
Thanks for the thoughts @duytnguyendtn, I think this is an important point. It seemed to me like the easiest way to get a functioning two-spectrum viewer using the jdaviz was to subclass `Specviz`, but I understand that the point of these configs is to provide _specific_ interfaces. I suppose a counterpoint would be to ask: do we expect a user who is visualizing spectra with a custom config to always create their own helper, which will likely have lots of [boiler-plate code](https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/specviz/helper.py#L28-L207)  common to most or every spectrum-related helper? 

> But doesn't the config file also specify plugins?
This is absolutely true, and I had forgotten that I built-in this limitation. I did choose which plugins were in the configuration so as to avoid the ones that would fail because the expected a viewer named `"spectrum-viewer"`. I'll see if I can get those working again now!

bmorris3 commented:
@duytnguyendtn – in [bc1b9b5](https://github.com/spacetelescope/jdaviz/pull/1681/commits/bc1b9b578718dcc4f43a0f892a229b1eae59fa57) I begin to address generalize the naming machinery across all available viz helpers (wow, that escalated quickly! 😅). I adjusted the test in `jdaviz/tests/test_app.py` to explicitly test that a subclass of the `Specviz` helper works as intended, providing access to viewers with nonstandard names.

pllim commented:
tl;dr -- Where do we go from here? Brett, did you test adding/removing viewers in Imviz? It is important that workflow does not break with this patch.

Also, there are conflicts, so please rebase when you are ready. Thanks!

bmorris3 commented:
@pllim I'm nearly done implementing @duytnguyendtn's suggested additional enhancement, which provides support for all plugins with custom viz helpers. I hope to have it done by the end of the day.

bmorris3 commented:
Ok, big changes in 288f764. I'll debrief everyone on Monday, but for now: now jdaviz will throw a warning when the configuration file tries to initialize a plugin that's incompatible with the viewers available to the viz helper. The plugin is skipped and the rest of the configuration is loaded.

bmorris3 commented on jdaviz/configs/default/plugins/gaussian_smooth/gaussian_smooth.py:
```diff
@@ -19,7 +19,13 @@
+        "_default_spectrum_viewer_reference_name":
+            ["spectrum_viewer_reference_name", {"require_spectrum_viewer": True}],
+    }
+)
```
Most recently in [cd3aee7](https://github.com/spacetelescope/jdaviz/pull/1681/commits/cd3aee73f412cd701cd69b7e9a79cb405ed76b0d) I address a comment from @kecnry via Slack that it'd be more sustainable to add the plugin constructor keyword arguments to the plugin machinery (rather than as a private dictionary within `app.py`). My solution is to add the requirements for the plugin to the `tray_registry` decorator. This way, future plugin authors can configure the requirements for their plugins.

bmorris3 commented on jdaviz/app.py:
```diff
@@ -1069,7 +1069,63 @@ def get_viewer_ids(self, prefix=None):
+                if require_no_selected_data and not len(viewer_item['selected_data_items']):
+                    return viewer_item['reference']
+                elif not require_no_selected_data:
+                    return viewer_item['reference']
```
This repeated block of code is a bit ugly, but I wasn't sure how best to produce this behavior without redundancy. Any tips on how to make this concise?

pllim commented on jdaviz/app.py:
```diff
@@ -1069,7 +1070,65 @@ def get_viewer_ids(self, prefix=None):
+                    elif not require_no_selected_data:
+                        return viewer_item['reference']
+            else:
+                if require_no_selected_data and not len(viewer_item['selected_data_items']):
+                    return viewer_item['reference']
+                elif not require_no_selected_data:
```
Does this not work?
```suggestion
                if ((require_no_selected_data and not len(viewer_item['selected_data_items']))
                        or (not require_no_selected_data)):
```

pllim reviewed (commented):
`+800 −244` 😱

I have not been following this one closely but I'll be happy to do a review when the existing discussions are settled. Feel free to ping me if my review is needed. Otherwise I'll leave this one in other devs' good hands.

bmorris3 commented:
@duytnguyendtn requested an overview of this PR now that it's ready for review. Here it comes!

## PR Overview

### Enhancements to existing infrastructure

* Added ``viz._default_*_viewer_reference_name`` class attributes to each helper, where ``*`` is one of: {``spectrum``, ``spectrum_2d``, ``flux``, ``image``, ``uncert``, ``table``}. Then I replace the arguments of most ``Application.get_viewer`` calls with the class attributes. The default behavior is preserved, so there are still some tests where these names persist.
* Added [``Application.get_first_viewer_reference_name``](https://github.com/spacetelescope/jdaviz/pull/1681/files#diff-33416aee3e2d4c0d86f96811f082119130c8598ef01ce1c9a7e852ecc39779b3R1074): get the first viewer's reference name which meets certain requirements. For example, when the kwarg ``require_spectrum_viewer=True`` the method will return the first available spectrum viewer.

### Tray item enhancements
* Added optional kwargs to the [``@tray_registry``](https://github.com/spacetelescope/jdaviz/pull/1681/files#diff-44f616436de8315fb25244c8555ea83596e03bedf23685221f4becc69bad8fbbR26-R34) decorator which specify the class attributes expected for each viewer reference name required by this tray item
* Added corresponding kwargs to the [``TrayRegistry.[add, __call__]``](https://github.com/spacetelescope/jdaviz/pull/1681/files#diff-be9456050fcc6fbeadd225a1aee9471fb737a070c8dee738f9f20e3818360a2f) methods (to enable the behavior described above) 
* When [initializing tray items](https://github.com/spacetelescope/jdaviz/pull/1681/files#diff-33416aee3e2d4c0d86f96811f082119130c8598ef01ce1c9a7e852ecc39779b3R1722-R1739) within ``Application``, supply ``viewer_reference_name_kwargs`` supplied by the `@tray_registry` decorators, which specify the viewer reference names required by each tray item.

### Points for discussion

1. ``Application.load_data``

   Throughout the tests (and plausibly throughout people's existing codes, too!), the generic ``Application.load_data`` method was used to load observations into the viewers. Since the ``_default_*_viewer_reference_name`` class attributes are defined on the viz helpers, the ``Application.load_data`` method is agnostic to the viewers available within the application. This means that unmodified calls to ``Application.load_data`` will fail unless they also specify the reference name for the viewer into which they should load the data. As a result, in many places (see examples within [model fitting plugin](https://github.com/spacetelescope/jdaviz/pull/1681/files#diff-fe0b4be38b96612e57a26980f1fd580fd574ec8fddd1daa08bb21f77fec5fe56), for example) I changed calls to ``viz.load_data`` to the viz-specific load methods like ``specviz.load_spectrum``. This way the viz helper specifies the correct viewer reference names for the type of observations that are being loaded.
   <br />

2. ``TrayRegistry``

   The new ``viewer_reference_name_kwargs`` dictionary that users must supply to the ``@tray_registry`` decorator is rather unintuitive. I documented how to use it [here](https://github.com/spacetelescope/jdaviz/pull/1681/files#diff-be9456050fcc6fbeadd225a1aee9471fb737a070c8dee738f9f20e3818360a2fR121-R140), but I'm wondering if we can do better. One improvement option could be to add a new (util?) function which defines viewer reference names and their requirements. Then the user could call the tray registry decorator with something like: ``viewer_reference_name_kwargs=required_reference_names(['spectrum', 'spectrum-2d', ...])``, and the function will return the formatted dictionary. Alternatively, and perhaps closer to the precedents within jdaviz, we could make a new ``ReferenceNameRegistry`` class which provides a decorator for functions which are also decorated by ``TrayRegistry``. Then the functions that call ``@tray_registry`` can also call ``@reference_name_registry(['spectrum', 'spectrum-2d', ...])``.  I thought I'd ask for your input on this before I added to our already highly-registered package.
   <br />

3. ``viz._default_*_viewer_reference_name``

   The ``"default"`` within the class attribute isn't strictly necessary. Do we like it?

duytnguyendtn commented on jdaviz/configs/cubeviz/helper.py:
```diff
@@ -14,14 +14,18 @@
+    _default_spectrum_viewer_reference_name = "spectrum-viewer"
+    _default_uncert_viewer_reference_name = "uncert-viewer"
+    _default_flux_viewer_reference_name = "flux-viewer"
+    _default_image_viewer_reference_name = "image-viewer"
```
What is this image viewer for in cubeviz?

duytnguyendtn commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -37,18 +38,37 @@ def __init__(self, *args, **kwargs):
+    def load_spectrum(self, data, data_label=None, format=None, show_in_viewer=True,
+                      spectrum_viewer_reference_name=None):
+
+        # If viewer reference name is not specified and
+        # the default viewer is available, use default
+        if spectrum_viewer_reference_name is None:
+            if (self._default_spectrum_viewer_reference_name in
+                    self.app.get_viewer_reference_names()):
+                spectrum_viewer_reference_name = self._default_spectrum_viewer_reference_name
+
+            # If viewer reference name is not specified and default is unavailable,
+            # use first spectrum viewer without loaded data:
+            else:
+                spectrum_viewer_reference_name = self.app.get_first_viewer_reference_name(
+                    require_spectrum_viewer=True, require_no_selected_data=True
+                )
+
```
I didn't notice this same check for the other configs; is this intentional?

duytnguyendtn commented on jdaviz/configs/specviz2d/helper.py:
```diff
@@ -132,6 +143,11 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
             Show data in viewer(s).
 
         """
+        kwargs.update({
+            "spectrum_2d_viewer_reference_name": self._default_spectrum_2d_viewer_reference_name,
+            "spectrum_viewer_reference_name": self._default_spectrum_viewer_reference_name,
+            "table_viewer_reference_name": self._default_table_viewer_reference_name
+        })
```
I've noticed this in the other helpers too: Why do we need to copy the viewer reference names into `kwargs` when we load data? I don't see the actual method using it, and if it's because we need to pass it to other methods, would it be more explicit to pass them in directly? If it's the latter, I'd feel this kind of obscures who needs it and why

duytnguyendtn commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -76,10 +78,12 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
     current_unit = None
-    current_spec = app.get_data_from_viewer("spectrum-viewer")
-    if current_spec != {} and current_spec is not None:
-        spec_key = list(current_spec.keys())[0]
-        current_unit = current_spec[spec_key].spectral_axis.unit
+    if spectrum_viewer_reference_name in app.get_viewer_reference_names():
+        current_spec = app.get_data_from_viewer(spectrum_viewer_reference_name)
+        if current_spec != {} and current_spec is not None:
+            spec_key = list(current_spec.keys())[0]
+            current_unit = current_spec[spec_key].spectral_axis.unit
+
```
Similar to https://github.com/spacetelescope/jdaviz/pull/1681#discussion_r990247038, I don't see this same check in the specviz2d helper. Struggling to see why this is only needed for specviz?

duytnguyendtn commented on jdaviz/configs/cubeviz/plugins/slice/slice.py:
```diff
@@ -14,7 +14,15 @@
 
 
-@tray_registry('cubeviz-slice', label="Slice")
+@tray_registry(
+    'cubeviz-slice', label="Slice",
+    viewer_reference_name_kwargs={
+        "_default_spectrum_viewer_reference_name":
+            ["spectrum_viewer_reference_name", {"require_spectrum_viewer": True}],
+        "_default_image_viewer_reference_name":
+            ["image_viewer_reference_name", {"require_image_viewer": True}],
+    }
+)
 class Slice(PluginTemplateMixin):
```
I like the idea of what this is trying to accomplish. I'm concerned it's a little opaque to remember exactly "where in the nested tree do I put `require_spectrum_viewer` again??"

Would it be possible to break it out to a separate decorator entirely to something like:

```
@tray_registry('cubeviz-slice', label="Slice")
@viewers_required('spectrum-viewer', 'image-viewer')
class Slice(PluginTemplateMixin):
```

I think something like the above would be easier from a code readability/maintainability standpoint

kecnry replied:
I really like this idea if it's feasible to migrate!  But since this is all internal, we could also postpone this to a future simplification

bmorris3 replied:
This is similar to my suggestion in https://github.com/spacetelescope/jdaviz/pull/1681#issuecomment-1271627712 (see Points for Discussion #2); I like this a lot too! @duytnguyendtn/@kecnry: let me know if you'd like to see that in this PR or a separate one.

duytnguyendtn replied:
I would love to see it in this PR, but I can also understand the desire not let this languish. I'll leave it up to you @bmorris3 if you think it's reasonable to implement now, or if we run the risk with complicated merge conflicts by letting this stale on for long

bmorris3 replied:
I can try to put it together in the next ~hour, and if it doesn't pan out I'll update you in this thread. 👍🏻

duytnguyendtn commented on jdaviz/core/template_mixin.py:
```diff
@@ -281,24 +281,42 @@ def hub(self):
+                 'id': viewer_item.get('id'),
+                 'icon': self.app.state.viewer_icons.get(viewer_item.get('id'))}
             if viewer_item.get('reference') is not None:
-                d['reference'] = viewer_item['reference']
-                d['label'] = viewer_item['reference']
+                d['reference'] = viewer_item.get('reference')
+                d['label'] = viewer_item.get('reference')
             else:
                 d['reference'] = None
-                d['label'] = viewer_item['id']
+                d['label'] = viewer_item.get('id')
```
No problems here, just curious; is this a stylistic decision or a code-based decision? As far as I know, `get` with no fallback behaves the same as `[]`

kecnry replied:
get with no fallback fallsback to a fallback of `None`

duytnguyendtn replied:
Ahh got it (I think... 😵). Thanks!

duytnguyendtn commented on CHANGES.rst:
```diff
@@ -27,6 +27,8 @@ New Features
 
 - Add Common Galactic line lists, split Atomic/Ionic list with verified medium info [#1656]
 
+- Add support for nonstandard viewer reference names [#1681]
```
It seems kind of... demeaning to reduce such a big infrastructural overhaul to just 7 words 🤣

duytnguyendtn commented on jdaviz/app.py:
```diff
@@ -1069,7 +1069,55 @@ def get_viewer_ids(self, prefix=None):
+        viewer has not yet loaded data, or e.g. ``require_spectrum_viewer``
+        to require that the viewer supports spectrum visualization.
+        """
+        from jdaviz.configs.specviz.plugins.viewers import SpecvizProfileView
+        from jdaviz.configs.specviz2d.plugins import SpectralExtraction
+        from jdaviz.configs.cubeviz.plugins.viewers import CubevizProfileView, CubevizImageView
+        from jdaviz.configs.mosviz.plugins.viewers import (
+            MosvizProfileView, MosvizTableViewer, MosvizProfile2DView
+        )
```
What are best practices for "inner" imports like this? I've always avoided since I've been unaware. I summon the arbiter @pllim!

pllim replied:
Ooops, just saw this. This delays the import until `_get_first_viewer_reference_name` is called, but I think that is called pretty soon anyway, so 🤷

pllim replied:
Actually, I think this mostly fixed non-GUI mode performance issues like #881 and #1362 ❗

duytnguyendtn commented on jdaviz/app.py:
```diff
@@ -1347,7 +1393,7 @@ def _update_selected_data_items(self, viewer_id, selected_items):
                 viewer.remove_data(data)
                 viewer._layers_with_defaults_applied= [layer_info for layer_info in viewer._layers_with_defaults_applied  # noqa
-                                                       if layer_info['data_label'] != data.label]
+                                                       if layer_info['data_label'] != data.label]  # noqa
```
There wasn't a codestyle fail here right? Is there a reason why a `# noqa` is required now?

pllim replied:
No reason, I think. Doesn't hurt but just confusing.

duytnguyendtn reviewed (approved):
Whew, that was a lot! Great work on this PR @bmorris3! What a way to get a tour of the codebase! I apologize for leaving all my review comments as single comments, but I figured this review would take some time and I figured I could get discussion going on some points while I reviewed the rest.

For completeness-sake, I didn't check that this PR touched all the places it *SHOULD* have (i.e. if there are other plugins that Brett missed), as that would take WAY too long IMHO. Though if we find any in the future, we should update them as we find them.

I still have a few questions I would prefer to have answered, but for the most part, those don't impact my approval of this PR. I also strongly advocate for https://github.com/spacetelescope/jdaviz/pull/1681#discussion_r990270715 to make this pill easier to swallow for the rest of the devs, but I leave that to @bmorris3 to decide whether to do now, or in a follow up.

This is really a dream come to fruition; I think this PR really puts `jdaviz` in the right direction to be a lot more flexible. Great work!

kecnry commented on jdaviz/app.py:
```diff
@@ -677,7 +676,8 @@ def get_data_from_viewer(self, viewer_reference, data_label=None,
 
         # If a data label was provided, return only the data requested
         if data_label is not None:
-            return data.get(data_label)
+            if data_label in data and len(data_label):
+                return data[data_label]
 
         return data
```
could this all be simplified to:

```
data.get(data_label, data)
```

kecnry commented on jdaviz/app.py:
```diff
@@ -1069,7 +1069,55 @@ def get_viewer_ids(self, prefix=None):
+            require_flux_viewer=False,
+            require_image_viewer=False
+    ):
+        """
+        Return the viewer reference name of the first available viewer.
+        Optionally use ``require_no_selected_data`` to require that the
+        viewer has not yet loaded data, or e.g. ``require_spectrum_viewer``
+        to require that the viewer supports spectrum visualization.
+        """
```
docstring could use expanding to list all parameters, especially since the method is not hidden with an underscore

kecnry commented on jdaviz/app.py:
```diff
@@ -1069,7 +1069,55 @@ def get_viewer_ids(self, prefix=None):
+        return [self._viewer_item_by_id(vid).get('reference') for vid in self._viewer_store]
+
+    def get_first_viewer_reference_name(
+            self, require_no_selected_data=False,
+            require_spectrum_viewer=False,
+            require_spectrum_2d_viewer=False,
+            require_table_viewer=False,
+            require_flux_viewer=False,
+            require_image_viewer=False
```
if multiple of these are passed as True, then they're not necessarily "required" anymore, since the first one that matches _any_ of the criteria is returned and dependent on the order the logic is hardcoded in the if statements), right?  I'm not sure if that means the logic should be changed, multiple True values should be forbidden, or if the input arguments should have a different prefix though...

Maybe for now we should just make this method hidden so we don't have to give it to much thought and can change it as we use it more in the future?

kecnry commented on jdaviz/app.py:
```diff
@@ -1673,7 +1719,24 @@ def compose_viewer_area(viewer_area_items):
+
+            optional_tray_kwargs = dict()
+
+            if tray_registry_options is not None:
```
Is it possible for `tray_registry_options` to be `None` (instead of an empty dict)?

kecnry commented on jdaviz/app.py:
```diff
@@ -1673,7 +1719,24 @@ def compose_viewer_area(viewer_area_items):
+            optional_tray_kwargs = dict()
+
+            if tray_registry_options is not None:
+                for opt_attr, [opt_kwarg, get_name_kwargs] in tray_registry_options.items():
+                    opt_value = getattr(
+                        self, opt_attr, self.get_first_viewer_reference_name(**get_name_kwargs)
+                    )
+
+                    if opt_value is None:
+                        continue
+
+                    optional_tray_kwargs[opt_kwarg] = opt_value
```
this block could use some comments for help following, especially later.

kecnry commented on jdaviz/configs/cubeviz/helper.py:
```diff
@@ -51,6 +55,12 @@ def load_data(self, data, data_label=None, **kwargs):
         if data_label:
             kwargs['data_label'] = data_label
+
+        kwargs.update({
+            "flux_viewer_reference_name": self._default_flux_viewer_reference_name,
+            "uncert_viewer_reference_name": self._default_uncert_viewer_reference_name,
+            "spectrum_viewer_reference_name": self._default_spectrum_viewer_reference_name,
+        })
```
Is there a reason why these passed here as `kwargs` instead of accessed directly in `load_data`?  If we want users to be able to override the values passed to `load_data`, then maybe we want `kwargs.setdefault` instead?

kecnry commented on jdaviz/configs/default/plugins/model_fitting/tests/test_plugin.py:
```diff
@@ -14,7 +14,7 @@
 
 def test_default_model_labels(specviz_helper, spectrum1d):
-    specviz_helper.load_data(spectrum1d)
+    specviz_helper.load_spectrum(spectrum1d)
```
does `load_data` skip some important logic introduced in this PR?  I'm wondering if we need to make it an alias to `load_spectrum` (and perhaps deprecate `load_spectrum`) since the distinction is quite confusing for the user (and for devs)... although that probably could/should be a follow-up PR.

kecnry commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -547,7 +559,7 @@ def load_metadata(self, data_obj, ids=None, spectra=False, sp1d=False):
                            parser_reference="mosviz-metadata-parser")
 
-    def load_1d_spectra(self, data_obj, data_labels=None):
+    def load_1d_spectra(self, data_obj, data_labels=None, **kwargs):
```
I don't love all of these accepting `**kwargs` as it makes it easy for typos to slip through unnoticed... is there any reasonable alternative?  I think this is related to a previous question I had about `load_data` just directly accessing `self._default_....reference_name` instead of having them passed through all the methods as `kwargs`.

kecnry commented on jdaviz/configs/mosviz/plugins/slit_overlay/slit_overlay.py:
```diff
@@ -97,10 +114,11 @@ def place_slit_overlay(self, *args, **kwargs):
 
-                if self.app.get_viewer("image-viewer").toolbar.active_tool is not None:
-                    self.app.get_viewer("image-viewer").toolbar.active_tool = None
+                if image_viewer.toolbar.active_tool is not None:
+                    image_viewer.toolbar.active_tool = None
```
out of scope since it isn't introduced here, but I'm curious why enabling the slit overlay disables any selected tool 🤔

kecnry commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -27,8 +28,9 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
     format : str
         Loader format specification used to indicate data format in
         `~specutils.Spectrum1D.read` io method.
+    viewer_reference_name : str
```
```suggestion
    spectrum_viewer_reference_name : str
```

kecnry commented on jdaviz/configs/specviz2d/helper.py:
```diff
@@ -13,14 +13,17 @@ class Specviz2d(ConfigHelper, LineListMixin):
     _default_configuration = "specviz2d"
+    _default_spectrum_viewer_reference_name = "spectrum-viewer"
+    _default_spectrum_2d_viewer_reference_name = "spectrum-2d-viewer"
+    _default_table_viewer_reference_name = "table-viewer"
```
specviz2d doesn't have a table viewer, by default

bmorris3 replied:
That's true. But when `specviz2d.load_data` is supplied with a value for `spectrum2d`, the default parser is `mosviz-spec2d-parser` invoked, which by default has `add_to_table==True`, so a table viewer is expected implicitly.

kecnry replied:
Ok, then that's probably fine for now, but might be a sign that we should consider refactoring some of the parsers down the road to avoid that.

kecnry commented on jdaviz/configs/specviz2d/helper.py:
```diff
@@ -178,9 +198,11 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
 
         else:
+            allowed_key = "spectrum_viewer_reference_name"
+            kwargs_for_spec1d = {allowed_key: kwargs[allowed_key]}
             self.app.load_data(spectrum_1d, data_label=spectrum_1d_label,
                                parser_reference="specviz-spectrum1d-parser",
-                               show_in_viewer=show_in_viewer)
+                               show_in_viewer=show_in_viewer, **kwargs_for_spec1d)
```
can this be simplified to:

```
self.app.load_data(..., spectrum_viewer_reference_name=kwargs['spectrum_viewer_reference_name'])
```

kecnry commented on jdaviz/core/registries.py:
```diff
@@ -109,16 +130,41 @@ def add(self, name, cls, label=None, icon=None):
             # objects (when determining if a specific plugin is open, for example)
+            viewer_reference_name_kwargs = {}
+
+            if viewer_requirements is not None:
```
could this whole if statement be avoided if `viewer_requirements` defaults to an empty list instead of `None`?

kecnry commented on jdaviz/core/template_mixin.py:
```diff
@@ -281,24 +281,42 @@ def hub(self):
                 if viewer.__class__.__name__ != 'MosvizTableViewer']
 
-    @cached_property
+    @property
```
does the caching logic fail because the viewer name might change during runtime? 😬

kecnry commented on jdaviz/core/template_mixin.py:
```diff
@@ -281,24 +281,42 @@ def hub(self):
+    @property
     def spectrum_viewer(self):
-        return self._plugin.app.get_viewer("spectrum-viewer")
+        if hasattr(self, '_default_spectrum_viewer_reference_name'):
+            viewer_reference = self._default_spectrum_viewer_reference_name
+        else:
+            viewer_reference = self.app.get_first_viewer_reference_name(
+                require_spectrum_viewer=True
+            )
+
+        return self._plugin.app.get_viewer(viewer_reference)
```
I wonder if this should be abstracted away in the app somewhere or be a part of the call to `get_first_viewer_reference_name(use_default=True)` or something similar?

Especially since this logic is then repeated at least once more below for `spectrum_2d_viewer`.

kecnry reviewed (commented):
Just a few questions on the code implementation.  I think as long as we keep these things behind private methods, we can always iterate on the details later as we have more cases that use the flexibility introduced here.  But I am a bit confused when to rely on the default names set during initialization vs having to rely on kwargs passed through methods.  Definitely like the revised plugin decorator syntax now though... that is a huge improvement to keep it simple for the plugin developer!

kecnry commented on jdaviz/app.py:
```diff
@@ -487,11 +486,6 @@ def load_data(self, file_obj, parser_reference=None, **kwargs):
         parser_reference : str or `None`
             The actual data parser to use. It must already be registered
             to glue's data parser registry. This is mainly for internal use.
-
-        **kwargs : dict
-            Additional keywords to be passed into the parser defined by
-            ``parser_reference``.
-
```
I think this used to exist on main and accidentally got deleted here when changing from kwargs to referencing the helper?

bmorris3 replied:
True! Fixing.
